design/01-proc-macro.md

```diff
@@ -0,0 +1,296 @@
+        - Scarb itself will not contain any Rust source code compilation capabilities.
+        - Scarb will require users to have Cargo available, if it spots a procedural macro that has not been yet compiled, i.e. there is not a shared object/DLL binary in Scarb's cache.
+        - Thanks to this, writing code not using dynamic procedural macros will not require having `rustc` installed.
+- The procedural macros will be compiled with stable ABI layout of structs passing the FFI border. This should guarantee Rust ABI compatibility, regardless of Cargo toolchain version available on user machine. The `dylib` crate type will be safe, and thus this should prevent runtime crashes.
+- Running Rust compiler, and storing `target` directory will be completely private thing to Scarb.
+    - Users will not be able to influence this.
+    - This will ensure procedural macro compilation will be as hermetic as possible.
+- Scarb will use [`libloading`](https://crates.io/crates/libloading) for loading built procedural macros shared objects.
+
+### Procedural macro API in Cairo plugins
+
+- The procedural macro have to be called from Cairo code in order to be executed during the compilation. In contrast to current behaviour of Cairo plugins, no longer will they be executed on each node of AST.
+- The procedural macro will be triggered by one of three Cairo expressions
+    - Macro call, e.g. `xyz!`
+    - Macro attribute, e.g. `#[xyz]`
+    - Macro derive, e.g. `#[derive(Xyz)]`
+- A Rust crate exposing an API for writing procedural macros for Cairo will be published for programmers. This crate will be automatically added to generated `Cargo.toml` file.
+- The procedural macro implementation will be a Rust function, accepting a `TokenStream` (described in detail in following sections) on input and returning the expansion result as an output.
+- The procedural macro implementation will return a `ProcMacroResult` defined in the helper library. The enum will contain one of three values:
+    - `ProcMacroResult::Leave` - procedural macro does not take any action
+    - `ProcMacroResult::Replace({ TokenStream, AuxData })` - procedural macro has generated code that replaces input
+    - `ProcMacroResult::Remove` - original code should be removed
+    - `ProcMacroResult::Bail(ProcMacroDiagnostics)` - procedural macro exited with an error
+- Alongside the new TokenStream, a procedural macro can emit auxiliary data, encoded as an arbitrary JSON.
+- The appropriate procedural macros will be executed based on the call in Cairo code by the new Cairo compiler internal `ProcMacroHost` plugin. This plugin will be called on each AST node and will decide if analyzed fragment requires code generation powered by an external plugin.
+
+**Example:**
+
+The procedural macro source code:
+
+```rust
+pub fn generate_code(token_stream: TokenStream) -> ProcMacroResult {
+	// Implementation skipped for brevity.
+}
+```
```

Comment:
Outdated. I strongly suggest making references to the guide instead of copy-pasting exactly because of this reason 😅